### DISCUSSION TITLE
PTZ Zones based on location preset

### DISCUSSION THREAD

**Discussion by docb117** [blakeblackshear/frigate#13599] [Ideas]
Since zones and PTZ camera's aren't fully integrated yet, how about:

Have frigate have a separate set of zones depending on what PTZ preset the camera is in.

For example, if I have a Garage PTZ camera (called Garage_PTZ), and I set up a preset with a view to the left (view_left) and  to the right (view_right).  In my instance, the left view has a different area of interest due to trees and other things.  So to have different zones for different presets, would the following config and supporting code changes work:

Example:  (coordinates are randomized )

Garage_PTZ
    .
    .
    zones:
      front_walk_zone:
         coordinates: 653,141,155,222,180,426,241,720,1122,720,1236,347,1237,0,817,0
         preset: left_view
     car_approach_zone:
        coordinates: 153,41,55,22,180,426,241,720,1122,720,236,347,237,0,817,0
         preset: right_view

**Comment by NickM-27** [comment]
> Have frigate have a separate set of zones depending on what PTZ preset the camera is in.

many cameras don't implement enough information to try this, and the ones that do don't actually tell you a preset that it is at so in most cases this would just not work well.

**Comment by hawkeye217** [comment]
We answered this recently for a reddit user here: https://www.reddit.com/r/frigate_nvr/comments/1f9rtd9/zones_that_support_ptz/

**Comment by docb117** [comment]
Well, here are two implementations that Frigate could include.  

One is easy, one is hard...maybe it's hard and very hard as I'm not a programmer.

For both implementations, Frigate is the primary source of a PTZ camera state.   

"Easy" implementation:
   When, through Frigate, a PTZ camera is moved to a  preset, update the zone/masking/etc
   ..Else, continue with normal Frigate...

"Hard" implementation:
  Again, relying on Frigate to be the source, use vision/object/?? detection to tag and then identify (visually) if the PTZ camera is at (near?) a preset.  If so, update zone/masking/etc
  ..Else, continue with normal Frigate...

**Comment by hawkeye217** [comment]
The onvif standard has a `GetStatus` call that, if supported by the camera, can return some aspects of the camera's position. However, not every camera supports this call - it seems like it just depends on how much the devs of the camera firmware invest in the onvif standard. Additionally, even if a camera does support `GetStatus`, there's nothing available in the onvif spec to be able to query if the current position is at a preset. 

Additionally, PTZ stepper motors have varying quality. A 1000 USD PTZ performs better than my 150 USD PTZ, which sometimes just can't put the camera back in the *exact* location it was in.

It's not impossible to write this code, but with varying support for the onvif standard and the varying quality of camera hardware, it would be impossible to support and be reliable for thousands of users.

So for all practical intents and purposes, yes, this is impossible, and as of right now, we do not plan to implement it.

**Comment by docb117** [comment]
The only thing Frigate needs to know from a PTZ camera is the name of a preset.  Anything else is either controlled by Frigate or ignored.   

All logic would be internal to Frigate.  

So under suggestion 1, if I moved a PTZ through a 3rd party app then nothing happens (current behavior).
If I told Frigate to move a PTZ camera to a preset, then Frigate will update the mask/zone/etc for the matching preset.

Suggestion 2 could work by checking the object recognition  tags after Frigate moves the camera.  
  If something else moves the camera, Frigate does nothing (current behavior) 

These suggestions don't rely on any information from the camera other than the name of the preset.  

I've also been graced with a work experience (think live televised enchantments for sports) that has even taught me there is a third way to implement this.  However, this falls under an extremely hard implementation.

**Reply by hawkeye217** [reply]
Of course keeping everything internal to Frigate is one possible option, but like I already mentioned, there are too many shortcomings for it to be accurate enough and supportable for all users, especially when Frigate can't reliably discern the actual position of the camera.

For the "extremely hard implementation" of point tracking, technically the 3rd party object tracker we already use (Norfair) supports image coordinate transformation (both through translation and homography). Frigate's autotracking already uses this for several of its features. But it has proven to be less than ideal in varying real-world conditions.

For now, this feature will have to be a pipe dream unless a new package comes out that implements a unique feature that makes it feasible, or the ONVIF standard is updated for cameras to report their current preset.

**Reply by NickM-27** [reply]
I'd also be curious to understand what this would actually be used for. I don't really see a point in having masks on a ptz camera besides any overlays because the camera is moving by design.

In the case of zones, the primary setup I'm familiar with is having stationary cameras that active the ptz to move to a certain preset, so the ptz having zones seems moot since the object already knowingly entered a zone of interest and the ptz is moving to the preset to get a better look

**Reply by docb117** [reply]
hawkeye217:  yes I understand how the camera's are pretty (very) dumb.  But even if the ONVIF standard is updated, not every camera will full (or correctly) implement it.  So the camera's will likely never be at the level of reporting their status that you are looking to implement with.   IMHO, the brains have to be fully in the local code as you can't trust the cameras.    Also, if it's not perfect that's fine with me.  


NickM-27:  As for use cases. here are a few:
Assume I have a house with a pool and a trampoline in the backyard.  
I also have Home Assistant (HA) automations and Frigate is integrated into HA.

In  the backyard, there is a BackyardPTZ with the following presets: PoolPreset, TrampolinePreset, BirdfeaderPreset
By default, BackyardPTZ watches TrampolinePreset and Frigate does it's job.
HA receives an alert from a pool occupancy sensor.  
This alert starts an automation (SafePoolAutomation) that has HA tell Frigate to move BackyardPTZ to PoolPreset 
SafePoolAutomation then lets Frigate record and send notifications based on the settings.  If after a while, the HA SafePoolAutomation stops receiving notifications (people have left the pool), then SafePoolAutomation tells frigate to move BackyardPTZ to TrampolinePreset and record and notify based on this different location.

I can come up with a different use case based on my front door too.  
Have the PTZ watch the front walkway until the doorbell is rung, then look at the doorbell area.

**Reply by NickM-27** [reply]
right, but none of these really require zones to work the way you describe. Especially if the camera is a zoom along with pan/tilt then it would likely only be viewing the area of interest anyway and the global camera sensors can be used for whatever automation logic is needed
